### PR TITLE
[REVIEW] Kafka list topics

### LINKED ISSUE #6131: [FEA] Add Kafka cluster metadata API to custreamz.kafka.
Confluent Kafka has an API for the consumer to get the metadata for a Kafka cluster. Here’s how I use it to get the number of partitions for a topic:

```
kafka_cluster_metadata = consumer.list_topics(topic)
npartitions = len(kafka_cluster_metadata.topics[topic].partitions)
```

In custreamz, we currently need to specify the number of partitions to start a stream, which is not ideal in case of production workloads where we need to add partitions to a topic, etc. No other streaming stacks need the end-user to specify #partitions for a stream. 

I can create a PR against streamz to use this `list_topics` API with CK, but it is currently lacking in the custreamz.kafka/cudf_kafka bit.

It would be great to port this API to custreamz.kafka. 

cc: @jdye64

### PR DESCRIPTION
**Pull request by jdye64**
Adding function to list all topics associated with the underlying Kafka broker connection. This includes the topic_names and all the the partitions (base 0 index of partition) associated with them.

This closes #6131

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: cpp/libcudf_kafka/src/kafka_consumer.cpp [hunk 1]
```diff
@@ -173,8 +193,6 @@ std::map<std::string, int64_t> kafka_consumer::get_watermark_offset(std::string
     if (err == RdKafka::ErrorCode::ERR__PARTITION_EOF) {
       results.insert(std::pair<std::string, int64_t>("low", low));
       results.insert(std::pair<std::string, int64_t>("high", high));
-    } else {
-      CUDF_FAIL("Error retrieving Kafka watermark offset from broker");
```
Comment (cwharris) [inline cpp/libcudf_kafka/src/kafka_consumer.cpp:177]: Is this related to the other changes?

Comment (jdye64) [reply cpp/libcudf_kafka/src/kafka_consumer.cpp:177]: @cwharris Thank you for catching this. This is a mistake and part of something else I am working on but seems to have mistakenly found its way in here. I will make another commit to replace it.

File: cpp/libcudf_kafka/src/kafka_consumer.cpp [hunk 2]
```diff
@@ -153,6 +153,26 @@ int64_t kafka_consumer::get_committed_offset(std::string const &topic, int parti
   return offset > 0 ? offset : 0;
 }
 
+const std::map<std::string, std::vector<int32_t>> kafka_consumer::list_topics()
+{
+  RdKafka::Metadata *md;
+  std::map<std::string, std::vector<int32_t>> topic_parts;
```
Comment (jsmaupin) [inline cpp/libcudf_kafka/src/kafka_consumer.cpp:159]: Are we returning just a list of partition integer values? It seems like the Confluent library returns so much more.  What does the result of the call to `consumer->metadata(...)` return? Do we need to parse the results? Would it make to not do any iterations here and just have this function be a pass-through for the `consumer-metadata(...)` function?

File: cpp/libcudf_kafka/include/cudf_kafka/kafka_consumer.hpp [hunk 3]
```diff
@@ -146,6 +146,13 @@ class kafka_consumer : public cudf::io::datasource {
    */
   int64_t get_committed_offset(std::string const &topic, int partition);
 
+  /**
+   * @brief Query the Kafka broker for the assigned topic Metadata
+   *
+   * @return Map of Kafka topics with their corresponding topic partition values.
+   */
+  const std::map<std::string, std::vector<int32_t>> list_topics();
```
Comment (kkraus14) [inline cpp/libcudf_kafka/include/cudf_kafka/kafka_consumer.hpp:154]: Higher level question: Is this something that we should add APIs for in libcudf / cudf Python or we should expect people to use `librdkafka` / `confluence-kafka-python` for?

File: python/custreamz/custreamz/kafka.py [hunk 4]
```diff
@@ -33,6 +33,14 @@ def __init__(self, kafka_configs):
 
         self.kafka_meta_client = KafkaDatasource(self.kafka_confs)
 
+    def list_topics(self):
+
+        """
+        List the topics associated with the underlying Kafka Broker connection.
```
Comment (jsmaupin) [inline python/custreamz/custreamz/kafka.py:39]: The Confluent library's `list_topic` function appears to return all metadata. Can we do the same?

Comment (jdye64) [reply python/custreamz/custreamz/kafka.py:39]: I think @kkraus14 makes a very valid point above ^^. When we spoke the other day that idea was to just send the partitions since all you really need is the len/number of partitions. However if you want *all* the metadata the best thing to do is just use python-confluent-kafka. There is no speedup and or need for us to add this to cudf_kafka. I think it best to just use python-confluent-kafka and us close this. Seems the real need is for something that behaves just like python-confluent-kafka so why not just us it?

File: cpp/libcudf_kafka/include/cudf_kafka/kafka_consumer.hpp (hunk 3 above)
Comment (jdye64) [reply cpp/libcudf_kafka/include/cudf_kafka/kafka_consumer.hpp:154]: @kkraus14 agreed. See https://github.com/rapidsai/cudf/pull/6220#discussion_r488084824

File: python/custreamz/custreamz/kafka.py (hunk 4 above)
Comment (chinmaychandak) [reply python/custreamz/custreamz/kafka.py:39]: > idea was to just send the partitions since all you really need is the len/number of partitions.

I think this should be OK. 

Using CK's `list_topics` isn't an option here, because the consumers created in streamz are the `custreamz.kafka` consumers. We shouldn't create an additional CK consumer just for this one call.  

If, however, we do end up needing all the metadata in the future, I agree — we should use CK's `list_topics` for sure. But it would definitely be in a different context to solve an entirely different problem, and I think there's no such problem we want to tackle in the short/medium-term roadmap.

Comment (chinmaychandak) [reply python/custreamz/custreamz/kafka.py:39]: The above answer is assuming that it is really difficult to have custreamz.kafka's `list_topics` to match CK's API. If it is just a little more work, then I think it is definitely worth doing like Jarod said.

File: cpp/libcudf_kafka/src/kafka_consumer.cpp [hunk 5]
```diff
@@ -153,6 +153,26 @@ int64_t kafka_consumer::get_committed_offset(std::string const &topic, int parti
...
+
+  const RdKafka::Metadata::TopicMetadataVector *mdv = md->topics();
+
+  for (auto &topic : *mdv) {
+    std::vector<int32_t> t_parts;
+    for (auto &partition : *(topic->partitions())) { t_parts.push_back(partition->id()); }
+    topic_parts[topic->topic()] = t_parts;
+  }
+
+  delete md, mdv;
```
Comment (harrism) [inline cpp/libcudf_kafka/src/kafka_consumer.cpp:172]: Pretty sure mdv is just a pointer to something owned by md. Are you sure it's correct to delete it?

**Comment by chinmaychandak** [comment]
@jdye64 Can we also add the retry for watermark offsets here?

**Comment by jsmaupin** [comment]
> @jdye64, can we also add the retry for watermark offsets to this PR?

Do we want to retry at this level? I think we already have a retry in Streamz correct?

**Comment by chinmaychandak** [comment]
> Do we want to retry at this level? I think we already have a retry in Streamz correct?

No, streamz doesn't [retry](https://github.com/python-streamz/streamz/blob/55e5fd79f57a82556c2f69291899281180282d77/streamz/sources.py#L508). Currently, with librdkafka, we skip every 15th batch because of this error. streamz doesn't retry, it just polls the offsets successfully on the 16th batch, as a result causing some lag. Ideally, we should dig into the root cause of why librdkafka errs to get the watermark offsets on every 15th batch, but if that's not possible, we need a retry at this level.

**Comment by jsmaupin** [comment]
> > Do we want to retry at this level? I think we already have a retry in Streamz correct?
> 
> No, streamz doesn't [retry](https://github.com/python-streamz/streamz/blob/55e5fd79f57a82556c2f69291899281180282d77/streamz/sources.py#L508). Currently, with librdkafka, we skip every 15th batch because of this error. streamz doesn't retry, it just polls the offsets successfully on the 16th batch, as a result causing some lag. Ideally, we should dig into the root cause of why librdkafka errs to get the watermark offsets on every 15th batch, but if that's not possible, we need a retry at this level.

That code looks like it does retry, though. It will try again when it loops back to the partition it skipped. I agree that skipping every 15th batch is a huge issue. If we have 15 partitions, then this is definitely a non-starter.  Are we proposing a work-around here for an issue we can root cause?

**Comment by jdye64** [comment]
> Do we want to retry at this level? I think we already have a retry in Streamz correct?

I definitely don't think we should be applying retry logic in the accelerated layer. If anything it should be handled higher up the stack is cudf_kafka, but I think the intention is more clear to the end user if it happens in custreamz.

File: cpp/libcudf_kafka/src/kafka_consumer.cpp (hunk 5 above)
Comment (jdye64) [reply cpp/libcudf_kafka/src/kafka_consumer.cpp:172]: Not entirely sure to be honest @harrism? I completely agree with and see your point but hte documentation makes it clear to call "delete" so to be honest. I'm not sure. Any ideas?

Comment (harrism) [reply cpp/libcudf_kafka/src/kafka_consumer.cpp:172]: I didn't know this was calling into an external library. If it says delete, do what it says I guess...  rdkafka API looks a bit old-school.

**Comment by jdye64** [comment]
rerun tests

**Comment by jdye64** [comment]
Waiting on an upstream change that is causing the tests to fail and then will upload again

**Comment by chinmaychandak** [comment]
@jdye64, are we waiting on more reviews here, or do we need to make some changes? I'm asking because this PR is unfortunately blocking a couple of other PRs on streamz.

**Comment by jdye64** [comment]
It is complete. Just more reviews

**Review by kkraus14 — APPROVED** [review]
Python/Cython LGTM

**Comment by chinmaychandak** [comment]
@vuule, can you please review when you get the time?

File: cpp/libcudf_kafka/src/kafka_consumer.cpp [hunk 6]
```diff
@@ -153,6 +153,26 @@ int64_t kafka_consumer::get_committed_offset(std::string const &topic, int parti
...
+{
+  RdKafka::Metadata *md;
+  std::map<std::string, std::vector<int32_t>> topic_parts;
+
+  CUDF_EXPECTS(RdKafka::ERR_NO_ERROR == consumer->metadata(true, nullptr, &md, default_timeout),
+               "Failed to list_topics in Kafka broker");
+
+  const RdKafka::Metadata::TopicMetadataVector *mdv = md->topics();
+
+  for (auto &topic : *mdv) {
```
Comment (vuule) [inline cpp/libcudf_kafka/src/kafka_consumer.cpp:166]: I think we don't really need the `mdv` variable, it's only used in the loop.
```suggestion
  for (auto const &topic : *std::unique_ptr<RdKafka::Metadata::TopicMetadataVector>{md->topics()}) {
```

File: cpp/libcudf_kafka/src/kafka_consumer.cpp [hunk 7]
```diff
@@ -153,6 +153,26 @@ int64_t kafka_consumer::get_committed_offset(std::string const &topic, int parti
   return offset > 0 ? offset : 0;
 }
 
+const std::map<std::string, std::vector<int32_t>> kafka_consumer::list_topics()
+{
+  RdKafka::Metadata *md;
+  std::map<std::string, std::vector<int32_t>> topic_parts;
+
+  CUDF_EXPECTS(RdKafka::ERR_NO_ERROR == consumer->metadata(true, nullptr, &md, default_timeout),
+               "Failed to list_topics in Kafka broker");
```
Comment (vuule) [inline cpp/libcudf_kafka/src/kafka_consumer.cpp:162]:
```suggestion
  auto const md = [&](){
    RdKafka::Metadata *md;
    CUDF_EXPECTS(RdKafka::ERR_NO_ERROR == consumer->metadata(true, nullptr, &md, default_timeout),
               "Failed to list_topics in Kafka broker");
    return std::unique_ptr<RdKafka::Metadata>{md};}();
  std::map<std::string, std::vector<int32_t>> topic_parts;
```

File: cpp/libcudf_kafka/src/kafka_consumer.cpp (hunk 5 above)
Comment (vuule) [inline cpp/libcudf_kafka/src/kafka_consumer.cpp:172]: you don't need these with other suggestions :)
```suggestion
```

**Review by vuule — CHANGES_REQUESTED** [review]
Some C++ suggestions, mostly to avoid the delete call and extra copies.

File: cpp/libcudf_kafka/src/kafka_consumer.cpp (hunk 6 above)
Comment (jdye64) [reply cpp/libcudf_kafka/src/kafka_consumer.cpp:166]: @vuule I like this but it doesn't compile ... ideas?

```/home/jdyer/Development/cudf/cpp/libcudf_kafka/src/kafka_consumer.cpp:167:78: error: no matching function for call to ‘std::unique_ptr<std::vector<const RdKafka::TopicMetadata*> >::unique_ptr(<brace-enclosed initializer list>)’
        *(std::unique_ptr<RdKafka::Metadata::TopicMetadataVector>{md->topics()})) {```

**Comment by jdye64** [comment]
@vuule does this look good to merge to you now?

**Review by vuule — APPROVED** [review]
Looks good, thanks!
